make track quality requirement in GeneralTracksImporter and PFTrackAl…

jsalfeld opened the pull request:
…goTools configurable

#### PR description:

Track Quality requirements in the PFBlockProducer are hardcoded. This PR is intended add configurability.

#### PR validation:

I ran OnLine_HLT_GRun.py with the new input for different track quality requirements in "process.hltParticleFlowBlock" PFBlockProducer. Locally I did a print out to check if bools are set correctly and the string correctly passed.

#### if this PR is a backport please specify the original PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

jsalfeld committed:
make track quality requirement in GeneralTracksImporter and PFTrackAlgoTools configurable

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12305

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc  modified in PR(s): #28110


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12305/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12305/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented on RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc:
```diff
@@ -31,10 +33,10 @@ class GeneralTracksImporter : public BlockElementImporterBase {
 
   edm::EDGetTokenT<reco::PFRecTrackCollection> src_;
   edm::EDGetTokenT<reco::MuonCollection> muons_;
+  std::string trackQuality_;
```
could you make this `const` ?

fwyzard commented on RecoParticleFlow/PFTracking/interface/PFTrackAlgoTools.h:
```diff
@@ -18,5 +18,6 @@ namespace PFTrackAlgoTools {
                         bool useIterTracking,
-                        bool debug = false);
+                        bool debug = false,
+			std::string trackQuality = "highPurity");
```
I would suggest to swap the order, and have `trackQuality` before `debug`.

makortel replied:
Also
```suggestion
			const std::string& trackQuality = "highPurity");
```

jsalfeld replied:
The change of order would complicate backward compatibility and introduce changes to where this function ("goodPtResolution") is used without trackQuality variable as input explicitly.

Let me know if this is ok (in principle I agree that having "debug" as last variable).

As far as I can tell, the affected classes are these ones:
RecoParticleFlow/PFTracking/plugins/HGCalTrackCollectionProducer.cc
RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc

jsalfeld replied:
@makortel is your suggested change ready to go into the commit or needs testing?
@fwyzard let me know if the other classes should be changed to accommodate the suggested change of order

makortel commented on RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc:
```diff
@@ -15,6 +15,8 @@ class GeneralTracksImporter : public BlockElementImporterBase {
         src_(sumes.consumes<reco::PFRecTrackCollection>(conf.getParameter<edm::InputTag>("source"))),
         muons_(sumes.consumes<reco::MuonCollection>(conf.getParameter<edm::InputTag>("muonSrc"))),
+	trackQuality_(
+	  (conf.existsAs<std::string>("trackQuality")) ? conf.getParameter<std::string>("trackQuality") : trackQuality_ = "highPurity"),
```
```suggestion
	  (conf.existsAs<std::string>("trackQuality")) ? conf.getParameter<std::string>("trackQuality") : "highPurity"),
```

makortel commented on RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc:
```diff
@@ -237,9 +237,11 @@ namespace PFTrackAlgoTools {
+			std::string trackQuality) {
+    
+    //check quality of tracks
+    if(!trackref->quality(reco::TrackBase::qualityByName(trackQuality)))
```
Actually, it would be better to do the string-to-enum conversion (`reco::TrackBase::qualityByName()`) once in the constructor (and possibly throw an exception if the returned value is `undefQuality`), and pass the enum to this function.

makortel replied:
Changing the `trackQuality` parameter here and `GeneralTracksImporter::trackQuality_` to `TrackBase::TrackQuality` enum? Yes, please do (and of course it needs to be eventually tested).

hatakeyamak commented:
Perhaps this is beyond the scope of this PR, but it may be nice if we can move these config parameter setting for PFBlockProducer to fillDescription-based, as I am doing for PFProducer on https://github.com/cms-sw/cmssw/pull/28110
(https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp)
which allow us to validate input parameters. Just wanted to mention this to see if this is something RECO conveners would prefer and Jakob would be willing to work on.

jsalfeld committed:
adjust according to suggestions, string enum in constructer, change order of debug and trackquality, no exception thrown though

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12339

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc  modified in PR(s): #28110
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc  modified in PR(s): #28110

cmsbuild commented:
A new Pull Request was created by  @jsalfeld (Jakob Salfeld-Nebgen) for master.

It involves the following packages:

RecoParticleFlow/PFProducer
RecoParticleFlow/PFTracking


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@mmarionncern, @makortel, @cbernet, @rovere, @lgray, @bachtis, @hatakeyamak, @seemasharmafnal this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7f08dd6630627b84ed6ef3a18386ad01d5cc4afd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ee0e7/3034/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ee0e7/3034/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2961064
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2960722
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

makortel commented on RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc:
```diff
@@ -15,6 +15,9 @@ class GeneralTracksImporter : public BlockElementImporterBase {
         muons_(sumes.consumes<reco::MuonCollection>(conf.getParameter<edm::InputTag>("muonSrc"))),
+        trackQuality_((conf.existsAs<std::string>("trackQuality"))
+                          ? reco::TrackBase::qualityByName(conf.getParameter<std::string>("trackQuality"))
+                          : reco::TrackBase::qualityByName("highPurity")),
```
Here and elsewhere, this could be just
```suggestion
                          : reco::TrackBase::highPurity),
```

jsalfeld committed:
simplify default quality flag (suggested by Matti)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12394

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc  modified in PR(s): #28110
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc  modified in PR(s): #28110

cmsbuild commented:
Pull request #28200 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented on RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc:
```diff
@@ -237,9 +237,10 @@ namespace PFTrackAlgoTools {
                         const std::vector<unsigned>& NHitCut,
                         bool useIterTracking,
+                        reco::TrackBase::TrackQuality trackQuality,
                         bool debug) {
```
I notice that the ```debug``` option is not needed any more by this method. It was introduced by commit https://github.com/cms-sw/cmssw/commit/5eafe51907989c8245f37f5fe46a7d8ce9a0c135#diff-05a1b01a408107e62fa1938501541abd  , but then made immediately after useless when moving all ```cout``` comment lines to ```LogDebug``` with https://github.com/cms-sw/cmssw/commit/4c64ac90711ec6a8eb9ac9e8c4a763035219887d#diff-05a1b01a408107e62fa1938501541abd

You can profit of this PR to remove that method parameter

jsalfeld replied:
@perotta done in commit below

perrotta commented on RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc:
```diff
@@ -123,7 +128,8 @@ void GeneralTracksImporterWithVeto::importToBlock(const edm::Event& e,
+            pftrackref->trackRef(), DPtovPtCut_, NHitCut_, useIterTracking_, trackQuality_, debug_)) {
       trkElem = new reco::PFBlockElementTrack(pftrackref);
       if (thisIsAPotentialMuon && debug_) {
         std::cout << "Potential Muon P " << pftrackref->trackRef()->p() << " pt " << pftrackref->trackRef()->p()
```
Also in this class the ```debug_``` parameter is not needed: just replace this line with a corresponding ```LogDebug``` (the same as already done in ```PFTrackAlgoTools.cc```) and remove all references to the ```debug_``` parameter in the code.

jsalfeld commented on RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc:
```diff
@@ -15,6 +15,9 @@ class GeneralTracksImporter : public BlockElementImporterBase {
       : BlockElementImporterBase(conf, sumes),
         src_(sumes.consumes<reco::PFRecTrackCollection>(conf.getParameter<edm::InputTag>("source"))),
         muons_(sumes.consumes<reco::MuonCollection>(conf.getParameter<edm::InputTag>("muonSrc"))),
+        trackQuality_((conf.existsAs<std::string>("trackQuality"))
```
@perrotta, @hatakeyamak, conf.existsAs is used throughout the PF code, may I suggest to move this proposal to another future PR?

hatakeyamak replied:
That's ok for me.

jsalfeld commented on RecoParticleFlow/PFTracking/interface/PFTrackAlgoTools.h:
```diff
@@ -18,5 +18,6 @@ namespace PFTrackAlgoTools {
                         const std::vector<double>& DPtovPtCut,
                         const std::vector<unsigned>& NHitCut,
                         bool useIterTracking,
+                        const reco::TrackBase::TrackQuality trackQuality = reco::TrackBase::highPurity,
```
I checked, that various inputs in the configuration are correctly converted into the proper enums.
I can remove the default value, since I included the default value in the input of the two other plugins making use of this function. Hence for the 3 plugins I adjusted, there should be no difference wrt the previous version.

jsalfeld committed:
removed debug parameter, removed default trackquality value from goodPtResolution method definition

cmsbuild commented:
The code-checks are being triggered in jenkins.

jsalfeld commented on RecoParticleFlow/PFTracking/plugins/HGCalTrackCollectionProducer.cc:
```diff
@@ -53,6 +53,7 @@ class HGCalTrackCollectionProducer : public edm::stream::EDProducer<> {
   // variables needed for copied goodPtResolution function
   // need to go back and figure out sensible values
   bool debug_;
```
@perotta done in commit below

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12413

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc  modified in PR(s): #28110
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc  modified in PR(s): #28110

cmsbuild commented:
Pull request #28200 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5a3dd48730c2b88d67d1304788801e26e605cd36
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ee0e7/3132/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ee0e7/3132/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2961036
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2960693
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

perrotta commented:
@jsalfeld shouldn't you update also some python config with the tunable "trackQuality" parameter?
Of course, for now that parameter will remain set to the default highPurity threshold, but it will be possible to configure it in some forthcoming update. Moreover, if the python config gets updated since now it will be picked in the next parsing of the HLT menu, and become available in there.

jsalfeld commented:
@fwyzard @Sam-Harper @gennai would you have a comment on @perrotta comment above?

hatakeyamak commented:
If I understand correctly, this new config parameter can be added to:
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py#L41-L50
but when it's not there, the default value is used, and I thought that's the behavior we want for now.

perrotta commented:
> If I understand correctly, this new config parameter can be added to:
> https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py#L41-L50
> but when it's not there, the default value is used, and I thought that's the behavior we want for now. So, PR is ok as it is on this point?

Yes, at the bare minimum I would add (already in this PR) the new parameter, with its default value set, in both 
- RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py
- RecoParticleFlow/PFTracking/python/hgcalTrackCollection_cfi.py

Moreover, if I am not wrong, the ```maxDPtOPt``` parameter in the blob that you pointed out is redundant (say useless), as it is correctly already defined in RecoParticleFlow/​PFProducer/​python/​particleFlow_cfi.py: if so, please remove it.

Then, the HGCalTrackCollectionProducer is simple enough that a ```fillDescriptions()``` method could be even added in this PR, thus being able to get rid of the ```existsAs()``` checks in there. I don't think this would be a complicate addition here, but I'm fine if you postpone it to a following PR.

Finally, the real completion of the work would be to implement ```fillDescriptions()``` in the PFBlockProducer as a whole. I agree that this is a more complex and delicate task that could be better postponed in a future and dedicated PR.

@Martin-Grunewald I don't understand how the parsing of the parameters of the PFBlockProducer is done. Shouldn't they be taken from the RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py?
If I look at the parameters in the HLT menu, I see that both the ```maxDPtOPt``` (and this is in principle ok, see above) and the ```cleanBadConvertedBrems``` (which should be set at its defalult "false" value, instead) are missing. So: where is the configuration for the HLT menu parsed from?

Sam-Harper commented:
> @fwyzard @Sam-Harper @gennai would you have a comment on @perrotta comment above?

If we wish to use any functionality in the HLT of the new features, my understanding is that we need a fillDescriptions() or at least the parameter set in the modules python file. Otherwise we cant access it and it would likely cause us problems. I understand the reluctance of the fillDescptions but could you not just at it to the python config? The one exception to this is if the parameter is part of a PSet although even in this case , it would be nice.

jsalfeld committed:
added (removed) trackQuality (debug) parameter from hgcalTrackCollection_cfi.py, added trackQuality parameter to particleFlowBlock_cfi.py

cmsbuild commented:
The code-checks are being triggered in jenkins.

jsalfeld commented:
I added the "trackQuality" parameter with default parameter to the files:
RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py
RecoParticleFlow/PFTracking/python/hgcalTrackCollection_cfi.py
From the latter I removed the redundant "debug" parameter, as was removed from the Producer in the previous commit.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28200/12465

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc  modified in PR(s): #28110
  - File  RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporterWithVeto.cc  modified in PR(s): #28110

cmsbuild commented:
Pull request #28200 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
> I added the "trackQuality" parameter with default parameter to the files:
> RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py
> RecoParticleFlow/PFTracking/python/hgcalTrackCollection_cfi.py
> From the latter I removed the redundant "debug" parameter, as was removed from the Producer in the previous commit.

Thank you @jsalfeld !
Please evaluate also the possibility of removing the ```maxDPtOPt``` parameter from the configuration of the particleFlowBlock prooducer. After that, all other requests that were listed in my previous [comment](https://github.com/cms-sw/cmssw/pull/28200#issuecomment-546675536) could even be postponed to some following PR

jsalfeld commented:
@perrotta, thanks a lot.

Let me know if I understand correctly: For maxDPtOPt:
> Please evaluate also the possibility of removing the maxDPtOPt parameter from the configuration of the particleFlowBlock prooducer.

My understanding is that the module python:

> RecoParticleFlow/​PFProducer/​python/​particleFlow_cfi.py

defines input for the PFProducer RecoParticleFlow/PFProducer/plugins/PFProducer.cc. Is it correct? As such they seem to be meant as being different. After digesting, I see that the PFProducer passes the parameters to PFAlgo.cc 
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/plugins/PFProducer.cc#L120

which in turn passes them to PFMuonAlgo.cc

https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/src/PFAlgo.cc#L26

Only here (for muons) the maxDPtOPt parameter is used. The same is true for other parameters like "trackQuality" and "minPixelHits"...

perrotta commented:
Indeed that's also my impression, Because of that, I guess that

perrotta commented:
Indeed that's also my impression. Because of it, I guess that

https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/python/particleFlow_cfi.py#L122

is the correct place where to define it, while in

https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py#L46

it is redundant and can be removed (just to clean up the config)

jsalfeld commented:
@perrotta , but I think
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py#L46

Is being used here (via GeneralTracksImporter and ultimately PFTrackAlgoTools.cc):
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFTracking/src/PFTrackAlgoTools.cc#L259

Whereas 
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/python/particleFlow_cfi.py#L122

is being used here (via PFProducer.cc and ultimately PFMuonAlgo.cc):
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/src/PFMuonAlgo.cc#L492

Isn't it?

perrotta commented:
Yes, but the DPtovPtCut_ in GeneralTracksImporter is taken from the configuration parameter "DPtOverPtCuts_byTrackAlgo", see https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/plugins/importers/GeneralTracksImporter.cc#L18.

There is no "maxDPtOPt" parameter ever read out from the config in GeneralTracksImporter

jsalfeld commented:
yes indeed, will do asap

jsalfeld committed:
remove redundant DPtOPt parameter in particleFlowBlock_cfi.py

cmsbuild commented:
The code-checks are being triggered in jenkins.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

hatakeyamak commented:
> yes indeed, will do asap. I assume these things will be fixed once the fillDescription is used and the cfi`s are produced automatically.

Thanks for sorting this out.

cmsbuild commented:
Comparison job queued.

hatakeyamak commented:
Comparisons look good to me.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2ee0e7/3202/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2961036
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2960693
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

perrotta commented:
+1
- The trackQuality to compare with is now made configurable in the mentioned classes
- Some cleaning of the code and python configurations is also implemented
- Jenkins tests pass and show no difference

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1

Martin-Grunewald commented:
@perrotta 
Just to confirm: PFBlockProducer is currently parsed from ```RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py```
